gufe/tests/test_protocol.py

```diff
@@ -20,43 +20,42 @@
     ProtocolUnitFailure,
 )
 
-from gufe.protocols.protocoldag import execute
+from gufe.protocols.protocoldag import execute_DAG
 
 from .test_tokenization import GufeTokenizableTestsMixin
 
 
 class InitializeUnit(ProtocolUnit):
-    @staticmethod
-    def _execute(ctx, *, settings, stateA, stateB, mapping, start, **inputs):
+    def _execute(self, ctx, **dependencies):
         return dict(
             log="initialized",
         )
 
 
 class SimulationUnit(ProtocolUnit):
-    @staticmethod
-    def _execute(ctx, *, initialization, **inputs):
+    def _execute(self, ctx, **dependencies):
 
-        output = [initialization.outputs['log']]
-        output.append("running_md_{}".format(inputs["window"]))
+        output = [dependencies['initialization'].outputs['log']]
+        output.append("running_md_{}".format(self._inputs["window"]))
 
         return dict(
             log=output,
-            window=inputs["window"],
-            key_result=(100 - (inputs["window"] - 10)**2),
+            window=self._inputs["window"],
+            key_result=(100 - (self._inputs["window"] - 10)**2),
             scratch=ctx.scratch,
             shared=ctx.shared
         )
 
 
 class FinishUnit(ProtocolUnit):
-    @staticmethod
-    def _execute(ctx, *, simulations, **inputs):
+    def _execute(self, ctx, **dependencies):
+        # sims are given a string of numeric key, extract them in ascending order
+        _, sims = zip(*sorted(dependencies.items(), key=lambda x: int(x[0])))
 
-        output = [s.outputs['log'] for s in simulations]
```

Comment:
Though this works, I think it introduces a bit too much dependency on naming conventions used when initializing a `ProtocolUnit`, such as `FinishUnit` here. Instead of a list of `SimulationUnit`s in `simulations`, now we have a dict with a very particular convention for names. This gets even more complicated if other `ProtocolUnit`s that aren't `SimulationUnit`s need to be included as dependencies.

Although we have no real examples so far of this, I believe the existing system makes it easier to write reusable `ProtocolUnit`s that are usable across many `Protocol`s. One could create a package of many `Protocol`s that use a set of `ProtocolUnit`s defined in the package as primitives, but this would get difficult to do if a given `ProtocolUnit` must take in a dict with a particular convention used for input instead of more Pythonic arguments.

Thoughts?